File: DSA/patterns/divide_conquer/mergeSort.js
function mergeSort(arr) {
  if (arr.length <= 1) return arr;

  let mid = Math.floor(arr.length / 2);
  let left = mergeSort(arr.slice(0, mid));
  let right = mergeSort(arr.slice(mid));

  return merge(left, right);
}

function merge(left, right) {
  let sortedArray = [];
  let i = 0,
    j = 0;

  while (i < left.length && j < right.length) {
    if (left[i] < right[j]) sortedArray.push(left[i++]);
    else sortedArray.push(right[j++]);
  }

  return [...sortedArray, ...left.slice(i), ...right.slice(j)];
}

// Example Usage
let arr = [6, 3, 8, 5, 2, 7, 4, 1];
console.log(mergeSort(arr)); // Output: [1, 2, 3, 4, 5, 6, 7, 8]

File: DSA/patterns/divide_conquer/merge_sort.py
#!/bin/python3

'''
    Implementation of Merge Sort.
'''

def merge_sort(arr):
    '''Sub Function.'''
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left_half = merge_sort(arr[:mid])
    right_half = merge_sort(arr[mid:])

    return merge(left_half, right_half)

def merge(left, right):
    '''Divide and Conquer Function.'''
    sorted_array = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            sorted_array.append(left[i])
            i += 1
        else:
            sorted_array.append(right[j])
            j += 1

    sorted_array.extend(left[i:])
    sorted_array.extend(right[j:])
    return sorted_array

# Example Usage
arr = [6, 3, 8, 5, 2, 7, 4, 1]
print(merge_sort(arr))  # Output: [1, 2, 3, 4, 5, 6, 7, 8]
